hypothesis-python/src/hypothesis/database.py

```diff
@@ -324,3 +335,287 @@ def delete(self, key: bytes, value: bytes) -> None:
+            if e.code == 401:
+                warning_message = (
+                    "Authorization failed when trying to download artifact from GitHub. "
+                    "Check your $GITHUB_TOKEN environment variable or make the repository public. "
+                )
+            else:
+                warning_message = (
+                    "Could not get the latest artifact from GitHub. "
+                    "This could be because because the repository or artifact does not exist. "
+                )
+        except URLError:
+            warning_message = "Could not connect to GitHub to get the latest artifact. "
+        except TimeoutError:
+            warning_message = "Could not connect to GitHub to get the latest artifact (connection timed out). "
+        if warning_message is not None:
+            warnings.warn(HypothesisWarning(warning_message))
+            return None
+
+        # Save the artifact to the cache
+        artifact_path = self.path / f"{datetime.datetime.now().isoformat()}.zip"
+        try:
+            with open(artifact_path, "wb") as f:
+                f.write(artifact_bytes)
+        except OSError:
+            warnings.warn(
+                HypothesisWarning("Could not save the latest artifact from GitHub. ")
+            )
+            return None
+
+        # Test that the artifact is valid
+        try:
+            with ZipFile(artifact_path) as f:
+                if f.testzip():
+                    raise BadZipFile
+        except BadZipFile:
+            warnings.warn(
+                HypothesisWarning(
+                    "The downloaded artifact from GitHub is invalid. "
+                    "This could be because the artifact was corrupted, "
+                    "or because the artifact was not created by Hypothesis. "
+                )
+            )
+            return None
+
+        return artifact_path
+
+    def _key_path(self, key) -> PurePath:
+        try:
+            return self.keypaths[key]
+        except KeyError:
+            pass
+
+        directory = PurePath(_hash(key) + "/")
+        self.keypaths[key] = directory
+        return directory
+
+    def fetch(self, key: bytes) -> Iterable[bytes]:
+        if self._disabled:
+            return
+
+        if not self._initialized:
+            self._initialize_db()
+            if self._disabled:
+                return
+
+        assert self._artifact is not None
+
+        kp = self._key_path(key)
+
+        with ZipFile(self._artifact) as zf:
+            # Check if kp exists
+            namelist = zf.namelist()
+            if kp not in (PurePath(p) for p in namelist):
+                return
+
+            # Iterate over files in kp
+            for filename in namelist:
+                fileinfo = zf.getinfo(filename)
+                if not fileinfo.is_dir() and filename.startswith(kp.as_posix()):
+                    with zf.open(filename) as f:
+                        yield f.read()
```

Comment:
I'm concerned that `O(n)` iteration might get expensive on large archives.  Can we cache a dictionary of `{dirname: set[filename]}` the first time we fetch?

We might also want to combine this with the integrity check I commented on above.